.github/workflows/check_crate.yml

```diff
@@ -0,0 +1,81 @@
+name: Check Crates
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+
+jobs:
+  check_crates:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v2
+
+      - name: Install dependencies
+        run: |
+          sudo apt-get update
+          sudo apt-get install -y jq
+
+      - name: Install Rust
+        uses: actions-rs/toolchain@v1
+        with:
+          toolchain: stable
+          profile: minimal
+          override: true
+
+      - name: Read member paths from main Cargo.toml
+        id: member_paths
+        run: |
+          paths=$(grep -oP '(?<=members = \[).*(?=\])' Cargo.toml | tr -d '\n')
+          paths=$(echo "$paths" | sed 's/,/\n/g' | tr -d '"')
+          echo "::set-output name=paths::$paths"
+
+      - name: Check for CHANGELOG.md
+        run: |
+          paths="${{ steps.member_paths.outputs.paths }}"
+          IFS=$'\n' read -r -d '' -a path_array <<< "$paths"
+          for path in "${path_array[@]}"; do
+            for crate in $(find $path -name 'Cargo.toml'); do
+              dir=$(dirname $crate)
+              if [ ! -f "$dir/CHANGELOG.md" ]; then
+                echo "Error: $dir/CHANGELOG.md is missing"
+                exit 1
+              fi
+            done
+          done
+
+      - name: Check for README.md
+        run: |
+          paths="${{ steps.member_paths.outputs.paths }}"
+          IFS=$'\n' read -r -d '' -a path_array <<< "$paths"
+          for path in "${path_array[@]}"; do
+            for crate in $(find $path -name 'Cargo.toml'); do
+              dir=$(dirname $crate)
+              if [ ! -f "$dir/README.md" ]; then
+                echo "Error: $dir/README.md is missing"
+                exit 1
+              fi
+            done
+          done
+
+      - name: Validate Cargo.toml categories
+        run: |
+          allowed_categories=$(curl -s https://crates.io/category_slugs | jq -r '.[]')
```

Comment:
https://crates.io/category_slugs doesn't support JSON, so we will need to manually get all the categries and hardcode them here
allowed_categories="cryptography networking ....."

Reply:
Done